src/main/java/org/glyptodon/guacamole/auth/json/connection/ConnectionService.java

```diff
@@ -0,0 +1,333 @@
+     *     established.
+     *
+     * @param info
+     *     Information associated with the connecting client.
+     *
+     * @return
+     *     A fully-established GuacamoleTunnel.
+     *
+     * @throws GuacamoleException
+     *     If an error occurs while connecting to guacd, or if permission to
+     *     connect is denied.
+     */
+    public GuacamoleTunnel connect(UserData.Connection connection,
+            GuacamoleClientInformation info) throws GuacamoleException {
+
+        // Retrieve proxy configuration from environment
+        GuacamoleProxyConfiguration proxyConfig = environment.getDefaultGuacamoleProxyConfiguration();
+
+        // Get guacd connection parameters
+        String hostname = proxyConfig.getHostname();
+        int port = proxyConfig.getPort();
+
+        // Generate and verify connection configuration
+        GuacamoleConfiguration config = getConfiguration(connection);
+        if (config == null) {
+            logger.debug("Configuration for connection could not be "
+                    + "generated. Perhaps the connection being joined is not "
+                    + "active?");
+            throw new GuacamoleResourceNotFoundException("No such connection");
+        }
+
+        // Determine socket type based on required encryption method
+        final ConfiguredGuacamoleSocket socket;
+        switch (proxyConfig.getEncryptionMethod()) {
+
+            // If guacd requires SSL, use it
+            case SSL:
+                socket = new ConfiguredGuacamoleSocket(
+                    new SSLGuacamoleSocket(hostname, port),
+                    config, info
+                );
+                break;
+
+            // Connect directly via TCP if encryption is not enabled
+            case NONE:
+                socket = new ConfiguredGuacamoleSocket(
+                    new InetGuacamoleSocket(hostname, port),
+                    config, info
+                );
+                break;
+
+            // Abort if encryption method is unknown
+            default:
+                throw new GuacamoleServerException("Unimplemented encryption method.");
+
+        }
+
+        final GuacamoleTunnel tunnel;
+
+        // If the current connection is not being tracked (no ID) just use a
+        // normal, non-tracking tunnel
+        final String id = connection.getId();
+        if (id == null)
+            tunnel = new SimpleGuacamoleTunnel(socket);
+
+        // Otherwise, create a tunnel with proper tracking which can be joined
+        else {
+
+            // Allow connection to be joined
+            final String connectionID = socket.getConnectionID();
+            final Collection<GuacamoleTunnel> existingTunnels = shadowers.putIfAbsent(connectionID,
+                    Collections.synchronizedList(new ArrayList<GuacamoleTunnel>()));
+
+            // Duplicate connection IDs cannot exist
+            assert(existingTunnels == null);
```

Comment:
I'm not sure I get what's going on here with this assert.. From the description: "however it is also documented behavior that connections which have the same ID will continue to function.", it sounds like having duplicate connection IDs is indeed something that can exist.

In the case of the activeConnections map right below, that behavior is explicitly checked for.

As far as I can tell, activeConnections and shadowers should always contain the same keys, since they are always added to and removed from together.

Under what situations would it be expected that activeConnections might have an ID that shadowers doesn't?

Reply:
There are two types of IDs in play: the value from the `id` property within the JSON (which is allowed to be duplicated) and the connection ID defined by the Guacamole protocol which is returned by guacd. The latter is guaranteed to be unique, and a duplicate would be a violation of the Guacamole protocol spec.

Reply:
Ah right, connectionId vs id. Not the same variable.

Reply:
To clarify here a bit further:

> Under what situations would it be expected that activeConnections might have an ID that shadowers doesn't?

Under all situations. These two maps actually use entirely different types of IDs as they keys. Compare `activeConnections` (which uses the arbitrary `id` value that *should* be unique but might not be):

https://github.com/glyptodon/guacamole-auth-json/blob/fc8a954bf419c741454591495ff09595fc8a6514/src/main/java/org/glyptodon/guacamole/auth/json/connection/ConnectionService.java#L97-L104

vs. `shadowers` (which uses Guacamole protocol connection IDs):

https://github.com/glyptodon/guacamole-auth-json/blob/fc8a954bf419c741454591495ff09595fc8a6514/src/main/java/org/glyptodon/guacamole/auth/json/connection/ConnectionService.java#L106-L112